probe-rs-cli-util/src/logging.rs

```diff
@@ -115,27 +192,179 @@ pub fn clear_progress_bar() {
     *guard = None;
 }
 
+fn send_logs() {
+    let mut log_guard = LOG.write().unwrap();
+
+    for breadcrumb in log_guard.drain(..) {
+        sentry::add_breadcrumb(breadcrumb);
+    }
+}
+
+fn sentry_config(release: String) -> sentry::ClientOptions {
+    sentry::ClientOptions {
+        dsn: Some(
+            Dsn::from_str("https://4396a23b463a46b8b3bfa883910333fe@sentry.technokrat.ch/7")
+                .unwrap(),
+        ),
+        release: Some(Cow::<'static>::Owned(release.to_string())),
+        #[cfg(debug_assertions)]
+        environment: Some(Cow::Borrowed("Development")),
+        #[cfg(not(debug_assertions))]
+        environment: Some(Cow::Borrowed("Production")),
+        default_integrations: false,
+        ..Default::default()
+    }
+}
+
+pub struct Metadata {
+    pub chip: Option<String>,
+    pub probe: Option<String>,
+    pub release: String,
+    pub commit: String,
+}
+
+/// Sets the metadata concerning the current probe-rs session on the sentry scope.
+fn set_metadata(metadata: &Metadata) {
+    sentry::configure_scope(|scope| {
+        metadata
+            .chip
+            .as_ref()
+            .map(|chip| scope.set_tag("chip", chip));
+        metadata
+            .probe
+            .as_ref()
+            .map(|probe| scope.set_tag("probe", probe));
+        scope.set_tag("commit", &metadata.commit);
+    })
+}
+
+fn print_uuid(uuid: Uuid) {
+    println(format!(
+        "  {} {} {}",
+        "Thank You!".cyan().bold(),
+        "Your error was reported successfully. If you don't mind, please open an issue on Github and include the UUID: ",
+        uuid
+    ));
+}
+
+/// Captures an std::error::Error with sentry and sends all previously captured logs.
+pub fn capture_error<E>(metadata: &Metadata, error: &E)
+where
+    E: Error + ?Sized,
+{
+    let _guard = sentry::init(sentry_config(metadata.release.clone()));
+    set_metadata(metadata);
+    send_logs();
+    let uuid = sentry::capture_error(error);
+    print_uuid(uuid);
+}
+
+/// Captures an anyhow error with sentry and sends all previously captured logs.
+pub fn capture_anyhow(metadata: &Metadata, error: &anyhow::Error) {
+    let _guard = sentry::init(sentry_config(metadata.release.clone()));
+    set_metadata(metadata);
+    send_logs();
+    let uuid = sentry::integrations::anyhow::capture_anyhow(error);
+    print_uuid(uuid);
+}
+
+/// Captures a panic with sentry and sends all previously captured logs.
+pub fn capture_panic(metadata: &Metadata, info: &PanicInfo<'_>) {
+    let _guard = sentry::init(sentry_config(metadata.release.clone()));
+    set_metadata(metadata);
+    send_logs();
+    let uuid = sentry::capture_event(PanicIntegration::new().event_from_panic_info(info));
+    print_uuid(uuid);
+}
+
+/// Ask for a line of text.
+fn text() -> std::io::Result<String> {
+    // Read up to the first newline or EOF.
+
+    let mut out = String::new();
+    std::io::stdin().read_line(&mut out)?;
```

Comment:
Does this work together with the console framework used in cargo-embed?

Reply:
Hmm good question ... I think so. I'll have to try, should be easy. I just simulated it outside of it :)